gluten-core/src/main/scala/org/apache/gluten/extension/columnar/OffloadSingleNode.scala

```diff
@@ -181,7 +185,127 @@ case class OffloadJoin() extends OffloadSingleNode with LogLevelUtil {
+    }
+  }
+
+  private def rewriteExpr(
+      expr: Expression,
+      replacedExprs: Map[String, AttributeReference]): Expression = {
+    expr match {
+      case _: InputFileName =>
+        replacedExprs.getOrElseUpdate(
+          expr.prettyName,
+          AttributeReference(expr.prettyName, StringType, false)())
+      case _: InputFileBlockStart =>
+        replacedExprs.getOrElseUpdate(
+          expr.prettyName,
+          AttributeReference(expr.prettyName, LongType, false)())
+      case _: InputFileBlockLength =>
+        replacedExprs.getOrElseUpdate(
+          expr.prettyName,
+          AttributeReference(expr.prettyName, LongType, false)())
+      case other =>
+        other.withNewChildren(other.children.map(child => rewriteExpr(child, replacedExprs)))
+    }
+  }
 
+  private def addMetadataCol(
+      plan: SparkPlan,
+      replacedExprs: Map[String, AttributeReference]): SparkPlan = {
+    def genNewOutput(output: Seq[Attribute]): Seq[Attribute] = {
+      var newOutput = output
+      for ((_, newAttr) <- replacedExprs) {
+        if (!newOutput.exists(attr => attr.exprId == newAttr.exprId)) {
+          newOutput = newOutput :+ newAttr
+        }
+      }
+      newOutput
+    }
+    def genNewProjectList(projectList: Seq[NamedExpression]): Seq[NamedExpression] = {
+      var newProjectList = projectList
+      for ((_, newAttr) <- replacedExprs) {
+        if (!newProjectList.exists(attr => attr.exprId == newAttr.exprId)) {
+          newProjectList = newProjectList :+ newAttr.toAttribute
+        }
+      }
+      newProjectList
+    }
+    plan match {
+      case f: FileSourceScanExec =>
+        f.copy(output = genNewOutput(f.output))
+      case f: FileSourceScanExecTransformer =>
+        f.copy(output = genNewOutput(f.output))
+      case b: BatchScanExec =>
+        b.copy(output = genNewOutput(b.output).asInstanceOf[Seq[AttributeReference]])
+      case b: BatchScanExecTransformer =>
+        b.copy(output = genNewOutput(b.output).asInstanceOf[Seq[AttributeReference]])
+      case p @ ProjectExec(projectList, child) =>
+        p.copy(genNewProjectList(projectList), addMetadataCol(child, replacedExprs))
+      case p @ ProjectExecTransformer(projectList, child) =>
+        p.copy(genNewProjectList(projectList), addMetadataCol(child, replacedExprs))
+      case _ => plan.withNewChildren(plan.children.map(addMetadataCol(_, replacedExprs)))
+    }
+  }
+
+  private def genProjectExec(projectExec: ProjectExec): SparkPlan = {
+    def findScanNodes(plan: SparkPlan): Seq[SparkPlan] = {
+      plan match {
+        case f @ (_: FileSourceScanExec | _: AbstractFileSourceScanExec |
+            _: DataSourceV2ScanExecBase) =>
+          Seq(f)
+        case o => o.children.flatMap(findScanNodes)
+      }
+    }
+
+    if (!TransformHints.isNotTransformable(projectExec)) {
+      logWarning(s"Columnar Processing for ${projectExec.getClass} is currently supported.")
```

Comment:
Why do we log a warning when it is supported?

Reply:
my misstake it only for local debug, will update